Feature/vosa 496 support for openjdk

skybert opened the pull request:
- support for OpenJDK 
- support for Oracle JDK is still in place
- turn off manifest JAR scanning (removes tons of errors)
- Integration tests for VBox machines

Commit:
Turn off Tomcat META-INF/MANIFEST.MF scanning

With a recent Tomcat, 9.0.13, loads of warnings about Tomcat being
unable to scan a JAR file which I never had reference nor did exist on
the file system:

```text
org.apache.tomcat.util.
scan.StandardJarScanner.scan Failed to scan
[file:/opt/tomcat-engine1/escenic/lib/serializer.jar] from classloader
hierarchy
```

It turned out, with Tomcat 8.0.41 or thereabouts, it has started to
scan all manifests for other JARs and add these to the classpath. If
any of these cannot be found, a big fat warning and stack trace is
added to the Tomcat standard out log.

For a typical Content Engine installation, this meant hundreds of
uninteresting log messages on each startup of the app server.

Commit:
Support for OpenJDK in ece-install and /usr/bin/ece

- configuration options ece-install.yaml:

  environment:
    java_vendor: "oracle"

- configuration options ece-install.conf:

   fai_java_vendor=oracle

- configuration options ece.conf and ece-<instance>.conf (this affects
  the JVM parameters):

   java_vendor=oracle

- OpenJDK is the default for both ece and ece-install and will
  currently install OpenJDK 11 on Debian, RedHat and their
  derivatives.

Commit:
Integration test using VirtualBox image (with snapshot)

- clean-slate integration test now reads ece-install-<distro>.yaml
- The clean-slate integration test now reads ece-install-<distro>.yaml
  and generates ece-install.yaml instead of ece-install-dev.yaml
- Ignore local var/cache/escenic

Commit:
Ensure Solr 6.5.1 can be used with OpenJDK 11

- Solr 6.5.x has an error in its test for Java version (it tests the
  ASCII value of the two version strings or something, hence version
  11.0.0 is deemed lower than 1.8

  For this reason, we need to massage the bin/solr command a bit. Also,
  we're removing any Oracle specific JVM option that's in there.

- when we upgrade to Solr 7.x, these hacks should not be necessary
  anymore as that branch of Solr (should) have proper support for OpenJDK
  1.8

Commit:
Only update the alternatives system if binary exists

Commit:
Ubuntu LTS 18.04 needs libyaml-libyaml-perl

- libyaml-perl isn't enough anymore
- this change is compatible with Debian stable

Commit:
Explicitly create /var/lib/escenic/engine

- This directory is normally created as a side effect, but sometimes,
  like on this fresh installation on Ubuntu LTS 18.04, it wasn't
  created, leading to the indexer not starting.

Commit:
No longer needed to pre-configure machine with Escenic APT repo

- Added APT credentials to sources.list entry as is already available
  to ece-install through:

  credentials:
    - site: apt.escenic.com
      user: APT_ESCENIC_USER
      password: APT_ESCENIC_PASSWORD

 - Now, as long as the user configures the correct code name for the
   release he/she wants, there's no need to add anything to
   /etc/apt/sources.list.d/escenic.list prior to running ece-install.

  environment:
    apt:
      escenic:
        pool: eyre

Commit:
Tell ece-install which APT code name to use (i.e. which CUE release)

Commit:
Feature: Possible to not install APR wrappers for Tomcat

- if the OS doesn't provide a compatible version of APR for the Tomcat
  installed by ece-install, you can now ask ece-install to not install
  APR by setting the following:

  environment:
    java_apr_install: false

  Default is 'true' (as before)

- example errors from incompatible Tomcat and APR wrappers:

  08-Jan-2019 13:53:39.602 SEVERE [main]
  org.apache.catalina.core.AprLifecycleListener.init An incompatible
  version [1.2.12] of the APR based Apache Tomcat Native library is
  installed, while Tomcat requires version [1.2.14]

modhassan commented:
Should we consider ubuntu 16.04?
```
E: Unable to locate package openjdk-11-jdk-headless
[ece-install-20] The command <apt-get install --no-install-recommends -o Dpkg::Options::=--force-confold --assume-yes openjdk-11-jdk-headless> run as user root ESC[37;31mESC[1mFAILEDESC[0m (the command exited with code 100), I'll exit now :-(
```
```
root@edemo-dev4:~# lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.3 LTS
Release:	16.04
Codename:	xenial
```
root@edemo-dev4:~# apt search openjdk-11*
Sorting... Done
Full Text Search... Done
```

skybert commented:
> Should we consider ubuntu 16.04?

To keep the task of maintaining `ece` and `ece-install` manageable, I'd say we only target the latest Debian stable, Ubuntu LTS, RHEL and CentOS. 

Machines running Ubuntu 16.04 have the several options to circumvent the problem you describe:
- use Oracle JDK (`.environment.java_vendor: oracle`) 
- install OpenJDK 8, 10 or 11 prior to running `ece-install`

mogsie commented on usr/local/src/integration-tests/virtual/clean-slate:
```diff
@@ -0,0 +1,128 @@
+      break
+    }
+    echo '.'
+    sleep 1
```
This might be a short time to sleep, I suggest `sleep $el`?

skybert replied:
Good idea

mogsie commented on usr/local/src/unit-tests/ece-install-java-test.sh:
```diff
@@ -15,6 +15,38 @@ test_can_get_oracle_rpm_url() {
 }
 
+test_can_get_java_spec_version() {
+  javac=$(which javac)
```
Is this testing the locally installed java?

skybert replied:
Yes. I should disable it, I suppose. Then you can enable it and adjust the path when you want to run it.

mogsie commented on usr/sbin/ece-install:
```diff
@@ -328,7 +328,8 @@ function install_common_os_packages()
       if ! grep -q apt.escenic.com /etc/apt/sources.list.d/escenic.list; then
         local package_pool=${fai_apt_vizrt_pool-stable}
-        add_apt_source "deb http://apt.escenic.com ${package_pool} main non-free"
+        add_apt_source "deb https://${fai_package_apt_user}:${fai_package_apt_password}@apt.escenic.com ${package_pool} main non-free"
```
Add the credentials to apt_auth.conf(5) instead:

```
echo machine apt.escenic.com \
     login "${fai_package_apt_user}" \
     password "${fai_package_apt_password}" \
     >> /etc/apt/auth.conf
```

mogsie commented on usr/share/escenic/ece-scripts/common-os.sh:
```diff
@@ -43,31 +43,20 @@ function get_user_home_directory() {
+  [[ "${java_vendor-openjdk}" == oracle ]]
+}
+
+function is_java_vendor_openjdk() {
```
This would return true (0) if java_vendor is unset. Is that important, or should it instead return false (1)?

skybert replied:
AFAICT, the functions work as expected (`openjdk` is the default):
```
function test_is_java_vendor_functions() {
  unset java_vendor
  printf "java_vendor: %s\n" "${java_vendor}"
  is_java_vendor_oracle
  printf "is_java_vendor_oracle %s\n" "${?}"
  is_java_vendor_openjdk
  printf "is_java_vendor_openjdk %s\n" "${?}"

  java_vendor=openjdk
  printf "java_vendor: %s\n" "${java_vendor}"
  is_java_vendor_oracle
  printf "is_java_vendor_oracle %s\n" "${?}"
  is_java_vendor_openjdk
  printf "is_java_vendor_openjdk %s\n" "${?}"

  java_vendor=oracle
  printf "java_vendor: %s\n" "${java_vendor}"
  is_java_vendor_oracle
  printf "is_java_vendor_oracle %s\n" "${?}"
  is_java_vendor_openjdk
  printf "is_java_vendor_openjdk %s\n" "${?}"
}
```

```
$ bash testing-tests.sh 
java_vendor: 
is_java_vendor_oracle 1
is_java_vendor_openjdk 0
java_vendor: openjdk
is_java_vendor_oracle 1
is_java_vendor_openjdk 0
java_vendor: oracle
is_java_vendor_oracle 0
is_java_vendor_openjdk 1
```

mogsie replied:
And that was sort of my point, that it returns true if java_vendor is not set, i.e. we're using a java version that we didn't recognise.  I'm saying that 

```
java_vendor: 
is_java_vendor_oracle 1
is_java_vendor_openjdk 0
```

is weird: both should return 1 (false) right?

skybert replied:
I disagree. Since `openjdk` is the default, `ece-install` should install `openjdk` if nothing is set. Think of it as a method to determine which product to install. (would it make more sense to you if it was named something like `is_java_vendor_openjdk_to_be_installed`?

In what use case would this implementation be a problem? Where there's a third vendor? Which would that likely be, GCJ?

modhassan commented:
Looks good to me.
Can ece-install consider available OpenJDK pakcage (e.g. openjdk-8) instead of openjdk-11-jdk-headless on Ubuntu 16.04?

skybert commented:
> Looks good to me.
> Can ece-install consider available OpenJDK pakcage (e.g. openjdk-8) instead of openjdk-11-jdk-headless on Ubuntu 16.04?

Let me see if I can make something that accommodates all Ubuntu and Debian releases without hard coding anything. ... it's not straight forward since there are significant changes between 8 and 11, but a best effort attempt by `ece-install` may go a long way and in some/most cases be enough. 

Update: I've pushed a fixup which should install whatever is the latest available openjdk headless package, see 894c2ec6893247bd3909ec4733473bf4d1657843. Does it float your boat, @modhassan ?

modhassan commented:
Yes. That's great!!
Now, `ece-install` does not even stop on Ubuntu 16.04 with `java_vendor: openjdk`!!!
But now I have noticed, It install `openjdk 10` on Ubuntu 18.04 though `openjdk 11` is available
```
root@ece-zebra:~# java -version
openjdk version "10.0.2" 2018-07-17
OpenJDK Runtime Environment (build 10.0.2+13-Ubuntu-1ubuntu0.18.04.4)
OpenJDK 64-Bit Server VM (build 10.0.2+13-Ubuntu-1ubuntu0.18.04.4, mixed mode)

```
```
root@ece-zebra:~# apt search openjdk-11-jdk*
Sorting... Done
Full Text Search... Done
default-jdk/bionic 2:1.10-63ubuntu1~02 amd64
  Standard Java or Java compatible Development Kit

default-jdk-headless/bionic 2:1.10-63ubuntu1~02 amd64
  Standard Java or Java compatible Development Kit (headless)

openjdk-11-jdk/bionic-updates,bionic-security 10.0.2+13-1ubuntu0.18.04.4 amd64
  OpenJDK Development Kit (JDK)

openjdk-11-jdk-headless/bionic-updates,bionic-security,now 10.0.2+13-1ubuntu0.18.04.4 amd64 [installed]
  OpenJDK Development Kit (JDK) (headless)
```

skybert commented:
> Yes. That's great!!

:smile: 

> Now, `ece-install` does not even stop on Ubuntu 16.04 with `java_vendor: openjdk`!!!

Awesome.

> But now I have noticed, It install `openjdk 10` on Ubuntu 18.04 though `openjdk 11` is available

`ece-install` does everything correctly, it's the `openjdk-11-jre-headless` package which is odd (out right wrong, I'd say):
```
root@ubuntu:~# /usr/lib/jvm/java-11-openjdk-amd64/bin/java -version
openjdk version "10.0.2" 2018-07-17
OpenJDK Runtime Environment (build 10.0.2+13-Ubuntu-1ubuntu0.18.04.4)
OpenJDK 64-Bit Server VM (build 10.0.2+13-Ubuntu-1ubuntu0.18.04.4, mixed mode)
root@ubuntu:~# dpkg -S /usr/lib/jvm/java-11-openjdk-amd64/bin/java
openjdk-11-jre-headless:amd64: /usr/lib/jvm/java-11-openjdk-amd64/bin/java
root@ubuntu:~# 
```

To verify which OpenJDK packages `ece-install` has installed, you can query `dpkg` as well as search the `.trail` file that `ece-install` leaves behind for each run:
```
root@ubuntu:~# dpkg -l openjdk-* | grep ^ii | awk '{print $2}'
openjdk-11-jdk-headless:amd64
openjdk-11-jre-headless:amd64
root@ubuntu:~# grep openjdk /var/lib/escenic/ece-install/1547713020-ece-install.trail 
trail_package_installed=openjdk-11-jdk-headless
root@ubuntu:~# 
```

modhassan commented:
Thanks for the answer :)

mogsie reviewed (commented):
Awesome...  Love / hate github's pending review feature...  Did you even see my comments? :)

mogsie commented on usr/share/escenic/ece-scripts/ece-install.d/java.sh:
```diff
@@ -32,7 +32,61 @@ function _java_get_oracle_rpm_url() {
+    local openjdk_package=
+    openjdk_package=$(java_find_latest_availale_openjdk_package)
+    if [ -z "${openjdk_package}" ]; then
+      openjdk_package=openjdk-11-jdk-headless
```
I would look into the option of installing `default-jdk-headless` here instead:  `Depends: default-jre-headless (= 2:1.11-68ubuntu1), openjdk-11-jdk-headless`

mogsie commented on usr/local/src/unit-tests/ece-install-java-test.sh:
```diff
@@ -15,6 +15,38 @@ test_can_get_oracle_rpm_url() {
+
+test_java_spec_10_is_ok() {
+  local expected=0
+  echo 11 | _java_is_spec_supported && actual=$? || actual=$?
```
Typo, should be `echo 10` I imagine.

mogsie commented on usr/share/escenic/ece-scripts/ece-install.d/content-engine.sh:
```diff
@@ -749,8 +749,8 @@ function install_ece_third_party_packages
 
-    if [ "$(has_oracle_java_installed)" -eq 0 ]; then
-      install_oracle_java
+    if ! has_java_installed; then
```
Sometimes we've seen that we have java installed, and we don't want java to be installed by ece-install, but the detection code doesn't work for some reason.  At the moment, if I install java 12 purposefully, I have no way of overriding ece-install to tell it not to install the "supported" Java 11, right?  I'd love to easily bypass the detection / installation of Java.

skybert replied:
You have always been allowed to skip installing Java if you've set:
```
environment:
  java_home: ${foo_java_home}
```
(or `java_home` in `ece-install.conf`)

I see now, though, that this only works for `java_vendor=oracle`. Let me fix that.

skybert commented:
> Awesome... Love / hate github's pending review feature... Did you even see my comments? :)

No. I was wondering why you didn't review it, but I figured you were swamped and I didn't want to nag you.

I've addressed your comments in https://github.com/escenic/ece-scripts/pull/71, @mogsie
